sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/MetricStorageUtils.java

```diff
@@ -20,6 +20,7 @@ private MetricStorageUtils() {}
    */
   static <T> void mergeInPlace(
       Map<Attributes, T> result, Map<Attributes, T> toMerge, Aggregator<T> aggregator) {
+    result.entrySet().removeIf(entry -> !toMerge.containsKey(entry.getKey()));
```

Comment:
This feels off to me. 

Is there supposed to be some hard guarantee that what we're merging in here should always have the same keyset?  This change feels like a different operation now -- more like `combineCommon` or something.

Reply:
This is actually part of what makes this simple solution possible. 

Currently, `DeltaMetricStorage` stores accumulations for synchronous instruments. It stores them as deltas, has no cap on the number of accumulations, and resets accumulations after each collection. When a cumulative collection happens, the delta accumulations are merged in the previous `Map<Attribute, T> result`.

So enforcing a max number of accumulations in `DeltaMetricStorage` alone doesn't actually bound memory usage - it only adds a bound for the number of accumulations in a single collection. This code forgets accumulations that weren't recorded in the most recent collection, and therefore allows the max accumulation enforcement in `DeltaMetricStorage` actually be the overall bound, not just the bound for a single collection. 

Arguably though, this code should be lifted up to `MetricStorageUtils`, so that this function have a single responsibility.

Reply:
I think this is a nice elegant solution.  Please update the javadoc to denote that you're limiting the output keyset to match the `toMerge` keyset.

I'd also be ok if you made a different function `mergeAndLimitKeysInPlace()` or some such.   Since these are just helpers for our own usage, I think it's find to have it here.


Same comment for `diffInPlace`

Reply:
Updated the javadoc, but left the method names the same. 👍